platforms/core-configuration/model-core/src/main/java/org/gradle/internal/reflect/validation/DefaultTypeAwareProblemBuilder.java

```diff
@@ -67,6 +73,67 @@ public TypeAwareProblemBuilder parentProperty(@Nullable String parentProperty) {
         return this;
     }
 
+    @Override
+    public Problem build() {
+        Problem problem = super.build();
+        String prefix = introductionFor(problem.getAdditionalData());
+        String text = Optional.ofNullable(problem.getContextualLabel()).orElseGet(() -> problem.getDefinition().getId().getDisplayName());
+        return problem.toBuilder().contextualLabel(prefix + text).build();
+    }
+
+    public static String introductionFor(Map<String, Object> additionalMetadata) {
+        StringBuilder builder = new StringBuilder();
+        String rootType = ofNullable(additionalMetadata.get(TYPE_NAME))
+            .map(Object::toString)
+            .filter(DefaultTypeAwareProblemBuilder::shouldRenderType)
+            .orElse(null);
+        DefaultPluginId pluginId = ofNullable(additionalMetadata.get(PLUGIN_ID))
+            .map(Object::toString)
+            .map(DefaultPluginId::new)
+            .orElse(null);
```

Comment:
:thinking: we should not use the `orElse` here and instead keep the `Optional` and use `isPresent` down the line.